### ISSUE TITLE
Allow inclusion of dependency in application local class loader when using dev tools

### ISSUE THREAD

**Issue by olivergierke** [spring-projects/spring-boot#4266]
When using DevTools, all dependencies of the project are loaded with a different class loader than the application code.

If the application consists of multiple modules, some of the dependencies might contain application components themselves. They might be package protected as they're wrapped by a publicly available component. If a JDK proxy has to be created for that package protected component (e.g. a Spring Data repository interface), the creation of the bean instance fails as the proxy creation enforces visibility rules and raises an exception.

It would be cool to be able to define a set of package names that should be loaded with the application class loader, too.

**Comment by cemo** [comment]
duplicates #3316 :)

**Comment by olivergierke** [comment]
I don't necessarily think it duplicates the ticket as the scenario here is slightly different and includes visibility of classes in terms of the scope, not necessarily the visibility from a certain classloader. So it's not – as in your example – a type from a JAR having to access a type from the project, but simply bootstrapping code in a JAR from that JAR classloader.

It's definitely related but not a duplicate, I think. Thanks for the pointer anyway.

**Comment by olivergierke** [comment]
I've prepared a minimal example in [this repository](https://github.com/olivergierke/spring-boot-devtools-sample). Follow the instructions in the readme to reproduce and workaround the issue.

**Comment by wilkinsona** [comment]
Thanks, Ollie.

This one's interesting (to me anyway). The proxy creation goes well to begin with. Proxy creation loads all four interfaces that are being proxied using the classloader that will define the proxy class. It checks the each interface that's loaded is the same (`==`) as the one being proxied. All of these checks pass which means that the `RestartClassLoader` is doing the right thing and delegating to its parent (the app class loader). Unfortunately, the wheels come off when `Proxy.defineClass0` is called:

```
Caused by: java.lang.IllegalAccessError: class example.lib.$Proxy31 cannot access its superinterface example.lib.LibRepository
	at java.lang.reflect.Proxy.defineClass0(Native Method)
	at java.lang.reflect.Proxy.access$300(Proxy.java:228)
	at java.lang.reflect.Proxy$ProxyClassFactory.apply(Proxy.java:642)
	at java.lang.reflect.Proxy$ProxyClassFactory.apply(Proxy.java:557)
	at java.lang.reflect.WeakCache$Factory.get(WeakCache.java:230)
	at java.lang.reflect.WeakCache.get(WeakCache.java:127)
	at java.lang.reflect.Proxy.getProxyClass0(Proxy.java:419)
	at java.lang.reflect.Proxy.newProxyInstance(Proxy.java:719)
	at org.springframework.aop.framework.JdkDynamicAopProxy.getProxy(JdkDynamicAopProxy.java:121)
	at org.springframework.aop.framework.ProxyFactory.getProxy(ProxyFactory.java:109)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport.getRepository(RepositoryFactorySupport.java:208)
	at org.springframework.data.repository.core.support.RepositoryFactoryBeanSupport.initAndReturn(RepositoryFactoryBeanSupport.java:251)
	at org.springframework.data.repository.core.support.RepositoryFactoryBeanSupport.afterPropertiesSet(RepositoryFactoryBeanSupport.java:237)
	at org.springframework.data.keyvalue.repository.support.KeyValueRepositoryFactoryBean.afterPropertiesSet(KeyValueRepositoryFactoryBean.java:71)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1637)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1574)
	... 33 more
```

That's a native method so I can't see exactly what's going on in there to trigger the problem. It's presumably because the `example.lib` package defined for the proxy class is different to the `example.lib` package that the super interfaces exist in. Unfortunately, overriding both `definePackage` methods on `RestartClassLoader` doesn't help as they're not called by whatever `defineClass0` is doing.

Long story short, this is another symptom of the underlying problem that we're tracking in #3316 where application classes are split across the restart class loader and the application class loader and things go south in a variety of ways as a result.
